### ISSUE TITLE
For-comprehensions with assignments break macro annotation expansions

### ISSUE THREAD

**Issue by lloydmeta** [scalameta/paradise#146]
Added a test [in my fork here](https://github.com/lloydmeta/paradise/blob/bug/for-compr/tests/annotations-meta/src/test/scala/compile/Expansion.scala#L6-L16) to demonstrate.

The simple test case demonstrates that for-comprehensions don't quite work well
with ScalaMeta annotations when there is an assignment (`=`) used in between generators
(`<-`).

```scala
for {
  n <- nums
  _ = n + 2
} yield n + 1
```

becomes (via `println(s.syntax)`)

```scala
nums.map(n => {
    val x$1 = n + 2
    (n, _)
  }).map((_: @scala.unchecked()) match {
    case (n, _) => n + 1
  })
```

Causing the following errors

```scala
[error] <macro>:5: missing parameter type for expanded function ((x$1) => scala.Tuple2(n, x$1))
[error]     (n, _)
[error]         ^
[error] <macro>:6: missing parameter type for expanded function ((x$2) => x$2: @scala.unchecked)
[error]   }).map((_: @scala.unchecked()) match {
[error]           ^
[error] <macro>:7: constructor cannot be instantiated to expected type;
[error]  found   : (T1, T2)
[error]  required: <error> => <error> @unchecked
[error]     case (n, _) => n + 1
[error]
```

**Comment by olafurpg** [comment]
Hmmm. This is a tricky one because we currently don't undo desugarings in for comprehensions. I suspect solving this issue properly involves solving for comprehensions, which requires some work. It's totally doable, but not a top priority at the moment. Until then, I'm afraid the best workaround is to either bind the `n + 2` expression to a bogus name or move the for comprehension into a separate file from the macro annotation.

**Comment by lloydmeta** [comment]
Yeah, it does seem tricky. I poked around to see if I could find a simple fix, but for now [my hack](https://github.com/lloydmeta/paradise/commit/569e12f3b32a188714f920bcbe6d47fafeb05fd5) involves disabling undesugaring (resugaring?) of synthetic names and lambdas, which "fixes" my test case, but causes 2 tests to fail, namely 

1. a.foreach(_ => bar())
2. a.foreach(_ + _)

I suppose this is to be expected because the for-comprehensions themselves are desugared by the compiler down to lambdas with synthetic names, which then gets mangled by the resugarings I disabled. This may mean that the easiest route is to have resugaring of for-comprehensions come before resugaring of lambdas and synthetic names. That way, the assumptions made for resugaring lambdas and synthetic names don't need to be revisited..

For completeness, here is the desugared AST of the above for-comprehension from the compiler:

```scala
nums.map(((n) => {
        val x$1 = n.$plus(2);
        scala.Tuple2(n, x$1)
      })).map(((x$2) => x$2: @scala.unchecked match {
        case scala.Tuple2((n @ _), _) => n.$plus(1)
      }))
```

**Comment by olafurpg** [comment]
Yes, I ran your example and stumbled into the same problem. Any workaround I could come up with will either be a temporary hack until for-comprehensions are properly handled or break other resugarings (I like that word 👍 ).

**Comment by lloydmeta** [comment]
I think [this issue in Scalameta](https://github.com/scalameta/scalameta/issues/148) may be related.

**Comment by DavidDudson** [comment]
Future conversation should be in the issue above. This issue will be closed when fixed in scalameta and merged into paradise.
